File: implementations/typescript-bun/src/planner.ts
import { normalizeConflictEdge, normalizeNode } from "./schema.ts";
import type {
  ConflictEdge,
  HeldPr,
  NormalizedPrNode,
  OrderingEdge,
  Plan,
  PlanningInput,
  PrNode,
  PrNumber,
  RebaseStep,
} from "./types.ts";

interface BatchPlan {
  readonly batches: readonly (readonly PrNumber[])[];
  readonly rebases: readonly RebaseStep[];
  readonly cycles: readonly PrNumber[];
}

const byNumber = (a: PrNode, b: PrNode): number => a.number - b.number;
const byEdge = (a: ConflictEdge, b: ConflictEdge): number => a.a - b.a || a.b - b.b;
const byOrdering = (a: OrderingEdge, b: OrderingEdge): number =>
  a.before - b.before || a.after - b.after;

export function buildPlan(input: PlanningInput): Plan {
  const nodes = input.nodes.map(normalizeNode).sort(byNumber);
  const conflicts = input.conflictEdges.map(normalizeConflictEdge).sort(byEdge);
  const ordering = buildOrdering(nodes, input.ancestryEdges);
  const eventual = planBatches(nodes, conflicts, ordering);
  const held = findHeld(nodes, ordering);
  const heldNumbers = new Set(held.map((entry) => entry.pr));
  const eligible = nodes.filter((node) => !heldNumbers.has(node.number));
  const eligibleNumbers = new Set(eligible.map((node) => node.number));
  const ready = planBatches(
    eligible,
    conflicts.filter((edge) => eligibleNumbers.has(edge.a) && eligibleNumbers.has(edge.b)),
    ordering.filter((edge) => eligibleNumbers.has(edge.before) && eligibleNumbers.has(edge.after)),
  );

  return {
    repository: input.repository,
    nodes: nodes.map(normalizeOutputNode),
    conflict_edges: conflicts,
    file_overlap_edges: findOverlaps(nodes),
    ordering_edges: ordering,
    stacks: buildStacks(ordering),
    suggested_landing_batches: eventual.batches,
    suggested_rebase_plan: eventual.rebases,
    ready_landing_batches: ready.batches,
    ready_now: ready.batches[0] ?? [],
    held_prs: held,
    ordering_cycles: eventual.cycles,
  };
}

function normalizeOutputNode(node: PrNode): NormalizedPrNode {
  return {
    pr: node.number,
    title: node.title,
    author: node.author ?? "unknown",
    head_ref: node.head_ref,
    base_ref: node.base_ref,
    draft: node.draft,
    mergeable: node.mergeable,
    review_decision: node.review_decision,
    additions: node.additions,
    deletions: node.deletions,
    files_count: new Set(node.files).size,
    base_conflict_paths: node.base_conflict_paths,
  };
}

function findOverlaps(nodes: readonly PrNode[]): ConflictEdge[] {
  const result: ConflictEdge[] = [];
  for (let leftIndex = 0; leftIndex < nodes.length; leftIndex += 1) {
    const left = nodes[leftIndex];
    if (left === undefined) continue;
    const leftFiles = new Set(left.files);
    for (let rightIndex = leftIndex + 1; rightIndex < nodes.length; rightIndex += 1) {
      const right = nodes[rightIndex];
      if (right === undefined) continue;
      const paths = right.files.filter((path) => leftFiles.has(path));
      if (paths.length > 0) result.push({ a: left.number, b: right.number, paths });
    }
  }
  return result;
}

function buildOrdering(
  nodes: readonly PrNode[],
  ancestryEdges: PlanningInput["ancestryEdges"],
): OrderingEdge[] {
  const byHead = new Map(nodes.map((node) => [node.head_ref, node] as const));
  const byPair = new Map<string, OrderingEdge>();
  for (const node of nodes) {
    const predecessor = byHead.get(node.base_ref);
    if (predecessor !== undefined && predecessor.number !== node.number) {
      const edge: OrderingEdge = {
        before: predecessor.number,
        after: node.number,
        reason: "base-ref",
      };
      byPair.set(`${edge.before}>${edge.after}`, edge);
    }
  }
  for (const edge of ancestryEdges) {
    const key = `${edge.before}>${edge.after}`;
    if (!byPair.has(key)) byPair.set(key, { ...edge, reason: "ancestry" });
  }
  return [...byPair.values()].sort(byOrdering);
}

function planBatches(
  nodes: readonly PrNode[],
  conflictEdges: readonly ConflictEdge[],
  orderingEdges: readonly OrderingEdge[],
): BatchPlan {
  const numbers = new Set(nodes.map((node) => node.number));
  const nodeByNumber = new Map(nodes.map((node) => [node.number, node] as const));
  const conflicts = setMap(numbers);
  const predecessors = setMap(numbers);
  const children = setMap(numbers);
  for (const edge of conflictEdges) {
    if (numbers.has(edge.a) && numbers.has(edge.b)) {
      required(conflicts, edge.a).add(edge.b);
      required(conflicts, edge.b).add(edge.a);
    }
  }
  for (const edge of orderingEdges) {
    if (numbers.has(edge.before) && numbers.has(edge.after)) {
      required(predecessors, edge.after).add(edge.before);
      required(children, edge.before).add(edge.after);
    }
  }

  const descendantCounts = new Map<PrNumber, number>();
  const descendantCount = (number: PrNumber): number => {
    const cached = descendantCounts.get(number);
    if (cached !== undefined) return cached;
    const reachable = new Set<PrNumber>();
    const pending = [...required(children, number)];
    while (pending.length > 0) {
      const child = pending.pop();
      if (child !== undefined && !reachable.has(child)) {
        reachable.add(child);
        pending.push(...required(children, child));
      }
    }
    descendantCounts.set(number, reachable.size);
    return reachable.size;
  };

  const remaining = new Set(numbers);
  const placed = new Set<PrNumber>();
  const batches: PrNumber[][] = [];
  let cycles: PrNumber[] = [];
  while (remaining.size > 0) {
    const ready = [...remaining].filter((number) => isSubset(required(predecessors, number), placed));
    if (ready.length === 0) {
      cycles = [...remaining].sort((a, b) => a - b);
      batches.push(...cycles.map((number) => [number]));
      break;
    }
    ready.sort((left, right) => {
      const leftNode = required(nodeByNumber, left);
      const rightNode = required(nodeByNumber, right);
      return (
        descendantCount(right) - descendantCount(left) ||
        countIntersection(required(conflicts, left), remaining) -
          countIntersection(required(conflicts, right), remaining) ||
        leftNode.additions + leftNode.deletions - (rightNode.additions + rightNode.deletions) ||
        compareText(leftNode.created_at, rightNode.created_at) ||
        left - right
      );
    });
    const batch: PrNumber[] = [];
    for (const number of ready) {
      if (batch.every((peer) => !required(conflicts, number).has(peer))) batch.push(number);
    }
    for (const number of batch) {
      remaining.delete(number);
      placed.add(number);
    }
    batches.push(batch);
  }

  const batchOf = new Map<PrNumber, number>();
  for (const [index, batch] of batches.entries()) {
    for (const number of batch) batchOf.set(number, index);
  }
  const rebases: RebaseStep[] = [];
  const rebaseOrder = [...numbers].sort(
    (left, right) => (batchOf.get(left) ?? 0) - (batchOf.get(right) ?? 0) || left - right,
  );
  for (const number of rebaseOrder) {
    const currentBatch = batchOf.get(number) ?? 0;
    const earlierConflicts = [...required(conflicts, number)]
      .filter((peer) => (batchOf.get(peer) ?? 0) < currentBatch)
      .sort((a, b) => a - b);
    const earlierDependencies = [...required(predecessors, number)]
      .filter((peer) => (batchOf.get(peer) ?? 0) < currentBatch)
      .sort((a, b) => a - b);
    const after = [...new Set([...earlierConflicts, ...earlierDependencies])].sort((a, b) => a - b);
    const reasons: RebaseStep["reasons"][number][] = [];
    if (earlierConflicts.length > 0) reasons.push("pair-conflict");
    if (earlierDependencies.length > 0) reasons.push("stack-dependency");
    if (after.length > 0) rebases.push({ pr: number, after, reasons });
  }
  return { batches, rebases, cycles };
}

function buildStacks(edges: readonly OrderingEdge[]): PrNumber[][] {
  const adjacency = new Map<PrNumber, Set<PrNumber>>();
  for (const edge of edges) addToSet(adjacency, edge.before, edge.after);
  const reduced = edges.filter(
    (edge) => !hasPath(adjacency, edge.before, edge.after, `${edge.before}>${edge.after}`),
  );
  const children = new Map<PrNumber, Set<PrNumber>>();
  const parents = new Map<PrNumber, Set<PrNumber>>();
  const involved = new Set<PrNumber>();
  for (const edge of reduced) {
    addToSet(children, edge.before, edge.after);
    addToSet(parents, edge.after, edge.before);
    involved.add(edge.before);
    involved.add(edge.after);
  }
  const stacks: PrNumber[][] = [];
  const visit = (node: PrNumber, path: readonly PrNumber[]): void => {
    const descendants = [...(children.get(node) ?? [])].sort((a, b) => a - b);
    if (descendants.length === 0) {
      if (path.length > 1) stacks.push([...path]);
      return;
    }
    for (const child of descendants) {
      if (!path.includes(child)) visit(child, [...path, child]);
    }
  };
  const roots = [...involved]
    .filter((number) => (parents.get(number)?.size ?? 0) === 0)
    .sort((a, b) => a - b);
  for (const root of roots) visit(root, [root]);
  return stacks;
}

function hasPath(
  adjacency: ReadonlyMap<PrNumber, ReadonlySet<PrNumber>>,
  start: PrNumber,
  target: PrNumber,
  skippedEdge: string,
): boolean {
  const pending = [start];
  const seen = new Set<PrNumber>();
  while (pending.length > 0) {
    const current = pending.pop();
    if (current === undefined || seen.has(current)) continue;
    seen.add(current);
    for (const child of adjacency.get(current) ?? []) {
      if (`${current}>${child}` === skippedEdge) continue;
      if (child === target) return true;
      pending.push(child);
    }
  }
  return false;
}

function findHeld(nodes: readonly PrNode[], ordering: readonly OrderingEdge[]): HeldPr[] {
  const held = new Map<PrNumber, string[]>();
  for (const node of nodes) {
    const reasons: string[] = [];
    if (node.draft) reasons.push("draft");
    if (node.base_conflict_paths.length > 0) reasons.push("local-base-conflict");
    if (node.mergeable === "CONFLICTING") reasons.push("github-base-conflicting");
    if (reasons.length > 0) held.set(node.number, reasons);
  }
  let changed = true;
  while (changed) {
    changed = false;
    for (const edge of ordering) {
      if (held.has(edge.before) && !held.has(edge.after)) {
        held.set(edge.after, [`depends-on-held:#${edge.before}`]);
        changed = true;
      }
    }
  }
  return [...held]
    .sort(([left], [right]) => left - right)
    .map(([pr, reasons]) => ({ pr, reasons }));
}

function setMap(numbers: ReadonlySet<PrNumber>): Map<PrNumber, Set<PrNumber>> {
  return new Map([...numbers].map((number): [PrNumber, Set<PrNumber>] => [number, new Set()]));
}

function addToSet<K, V>(map: Map<K, Set<V>>, key: K, value: V): void {
  const existing = map.get(key);
  if (existing === undefined) map.set(key, new Set([value]));
  else existing.add(value);
}

function countIntersection<T>(left: ReadonlySet<T>, right: ReadonlySet<T>): number {
  let count = 0;
  for (const value of left) if (right.has(value)) count += 1;
  return count;
}

function isSubset<T>(subset: ReadonlySet<T>, superset: ReadonlySet<T>): boolean {
  for (const value of subset) if (!superset.has(value)) return false;
  return true;
}

function required<K, V>(map: ReadonlyMap<K, V>, key: K): V {
  const value = map.get(key);
  if (value === undefined) throw new Error("internal planner invariant failed");
  return value;
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

File: implementations/python/pr_plan/planner.py
from __future__ import annotations

from pr_plan.model import (
    AnalysisInput,
    ConflictEdge,
    HeldPullRequest,
    OrderingEdge,
    Plan,
    PrNumber,
    PullRequest,
    RebaseEntry,
    RebaseReason,
)


def _ordering_edges(data: AnalysisInput) -> tuple[OrderingEdge, ...]:
    edges: dict[tuple[PrNumber, PrNumber], OrderingEdge] = {}
    by_head = {pr.head_ref: pr.number for pr in data.prs}
    for pr in data.prs:
        parent = by_head.get(pr.base_ref)
        if parent is not None and parent != pr.number:
            edges[(parent, pr.number)] = OrderingEdge(parent, pr.number, "base-ref")
    for before, after in data.ancestry_edges:
        edges.setdefault(
            (before, after), OrderingEdge(before, after, "ancestry")
        )
    return tuple(edge for _, edge in sorted(edges.items()))


def _file_overlaps(prs: tuple[PullRequest, ...]) -> tuple[ConflictEdge, ...]:
    file_sets = {pr.number: set(pr.files) for pr in prs}
    overlaps: list[ConflictEdge] = []
    for left_index, left in enumerate(prs):
        for right in prs[left_index + 1 :]:
            shared = tuple(sorted(file_sets[left.number] & file_sets[right.number]))
            if shared:
                overlaps.append(ConflictEdge(left.number, right.number, shared))
    return tuple(overlaps)


def _has_path(
    adjacency: dict[PrNumber, set[PrNumber]],
    start: PrNumber,
    target: PrNumber,
    skip: tuple[PrNumber, PrNumber],
) -> bool:
    pending = [start]
    seen: set[PrNumber] = set()
    while pending:
        current = pending.pop()
        if current in seen:
            continue
        seen.add(current)
        for child in adjacency.get(current, set()):
            if (current, child) == skip:
                continue
            if child == target:
                return True
            pending.append(child)
    return False


def _stacks(edges: tuple[OrderingEdge, ...]) -> tuple[tuple[PrNumber, ...], ...]:
    adjacency: dict[PrNumber, set[PrNumber]] = {}
    for edge in edges:
        adjacency.setdefault(edge.before, set()).add(edge.after)
    reduced = tuple(
        edge
        for edge in edges
        if not _has_path(
            adjacency, edge.before, edge.after, (edge.before, edge.after)
        )
    )
    children: dict[PrNumber, set[PrNumber]] = {}
    parents: dict[PrNumber, set[PrNumber]] = {}
    involved: set[PrNumber] = set()
    for edge in reduced:
        children.setdefault(edge.before, set()).add(edge.after)
        parents.setdefault(edge.after, set()).add(edge.before)
        involved.update((edge.before, edge.after))

    stacks: list[tuple[PrNumber, ...]] = []
    roots = sorted(number for number in involved if not parents.get(number))
    for root in roots:
        pending: list[tuple[PrNumber, tuple[PrNumber, ...]]] = [(root, (root,))]
        while pending:
            node, path = pending.pop()
            descendants = sorted(children.get(node, set()))
            if not descendants:
                if len(path) > 1:
                    stacks.append(path)
                continue
            for child in reversed(descendants):
                if child not in path:
                    pending.append((child, (*path, child)))
    return tuple(stacks)


def _held_prs(
    prs: tuple[PullRequest, ...], ordering: tuple[OrderingEdge, ...]
) -> tuple[HeldPullRequest, ...]:
    reasons: dict[PrNumber, list[str]] = {pr.number: [] for pr in prs}
    for pr in prs:
        if pr.draft:
            reasons[pr.number].append("draft")
        if pr.base_conflict_paths:
            reasons[pr.number].append("local-base-conflict")
        if pr.mergeable == "CONFLICTING":
            reasons[pr.number].append("github-base-conflicting")

    changed = True
    while changed:
        changed = False
        for edge in ordering:
            if reasons[edge.before] and not reasons[edge.after]:
                reasons[edge.after] = [f"depends-on-held:#{edge.before}"]
                changed = True
    return tuple(
        HeldPullRequest(number, tuple(pr_reasons))
        for number, pr_reasons in sorted(reasons.items())
        if pr_reasons
    )


def _landing_batches(
    prs: tuple[PullRequest, ...],
    ordering: tuple[OrderingEdge, ...],
    conflicts: tuple[ConflictEdge, ...],
) -> tuple[tuple[tuple[PrNumber, ...], ...], tuple[PrNumber, ...]]:
    remaining = {pr.number for pr in prs}
    if not remaining:
        return (), ()
    by_number = {pr.number: pr for pr in prs}
    conflicts_by_pr: dict[PrNumber, set[PrNumber]] = {
        number: set() for number in remaining
    }
    predecessors: dict[PrNumber, set[PrNumber]] = {
        number: set() for number in remaining
    }
    children: dict[PrNumber, set[PrNumber]] = {
        number: set() for number in remaining
    }
    for conflict_edge in conflicts:
        if conflict_edge.a in remaining and conflict_edge.b in remaining:
            conflicts_by_pr[conflict_edge.a].add(conflict_edge.b)
            conflicts_by_pr[conflict_edge.b].add(conflict_edge.a)
    for ordering_edge in ordering:
        if ordering_edge.before in remaining and ordering_edge.after in remaining:
            predecessors[ordering_edge.after].add(ordering_edge.before)
            children[ordering_edge.before].add(ordering_edge.after)

    descendant_cache: dict[PrNumber, int] = {}

    def descendant_count(number: PrNumber) -> int:
        cached = descendant_cache.get(number)
        if cached is not None:
            return cached
        reachable: set[PrNumber] = set()
        pending = list(children[number])
        while pending:
            child = pending.pop()
            if child not in reachable:
                reachable.add(child)
                pending.extend(children[child])
        result = len(reachable)
        descendant_cache[number] = result
        return result

    batches: list[tuple[PrNumber, ...]] = []
    placed: set[PrNumber] = set()
    cycle_nodes: tuple[PrNumber, ...] = ()
    while remaining:
        available = [
            number for number in remaining if predecessors[number].issubset(placed)
        ]
        if not available:
            cycle_nodes = tuple(sorted(remaining))
            batches.extend((number,) for number in cycle_nodes)
            break
        available.sort(
            key=lambda number: (
                -descendant_count(number),
                len(conflicts_by_pr[number] & remaining),
                by_number[number].additions + by_number[number].deletions,
                by_number[number].created_at,
                number,
            )
        )
        batch: list[PrNumber] = []
        for candidate in available:
            if all(selected not in conflicts_by_pr[candidate] for selected in batch):
                batch.append(candidate)
        batches.append(tuple(batch))
        remaining.difference_update(batch)
        placed.update(batch)
    return tuple(batches), cycle_nodes


def _rebase_plan(
    batches: tuple[tuple[PrNumber, ...], ...],
    ordering: tuple[OrderingEdge, ...],
    conflicts: tuple[ConflictEdge, ...],
) -> tuple[RebaseEntry, ...]:
    batch_of = {
        number: batch_index
        for batch_index, batch in enumerate(batches)
        for number in batch
    }
    dependencies: dict[PrNumber, set[PrNumber]] = {}
    reasons: dict[PrNumber, set[RebaseReason]] = {}
    for ordering_edge in ordering:
        if batch_of.get(ordering_edge.before, -1) < batch_of.get(
            ordering_edge.after, -1
        ):
            dependencies.setdefault(ordering_edge.after, set()).add(
                ordering_edge.before
            )
            reasons.setdefault(ordering_edge.after, set()).add("stack-dependency")
    for conflict_edge in conflicts:
        a_batch = batch_of.get(conflict_edge.a)
        b_batch = batch_of.get(conflict_edge.b)
        if a_batch is None or b_batch is None or a_batch == b_batch:
            continue
        earlier, later = (
            (conflict_edge.a, conflict_edge.b)
            if a_batch < b_batch
            else (conflict_edge.b, conflict_edge.a)
        )
        dependencies.setdefault(later, set()).add(earlier)
        reasons.setdefault(later, set()).add("pair-conflict")

    reason_order: dict[RebaseReason, int] = {
        "pair-conflict": 0,
        "stack-dependency": 1,
    }
    return tuple(
        RebaseEntry(
            pr=number,
            after=tuple(sorted(after)),
            reasons=tuple(sorted(reasons[number], key=reason_order.__getitem__)),
        )
        for number, after in sorted(
            dependencies.items(), key=lambda item: (batch_of[item[0]], item[0])
        )
    )


def make_plan(data: AnalysisInput) -> Plan:
    ordering = _ordering_edges(data)
    conflicts = tuple(
        sorted(data.conflict_edges, key=lambda edge: (edge.a, edge.b, edge.paths))
    )
    held = _held_prs(data.prs, ordering)
    held_numbers = {item.pr for item in held}
    suggested_batches, ordering_cycles = _landing_batches(data.prs, ordering, conflicts)
    ready_prs = tuple(pr for pr in data.prs if pr.number not in held_numbers)
    ready_batches, _ = _landing_batches(ready_prs, ordering, conflicts)
    return Plan(
        repository=data.repository,
        nodes=data.prs,
        conflict_edges=conflicts,
        file_overlap_edges=_file_overlaps(data.prs),
        ordering_edges=ordering,
        stacks=_stacks(ordering),
        suggested_landing_batches=suggested_batches,
        suggested_rebase_plan=_rebase_plan(suggested_batches, ordering, conflicts),
        ready_landing_batches=ready_batches,
        ready_now=ready_batches[0] if ready_batches else (),
        held_prs=held,
        ordering_cycles=ordering_cycles,
    )
